### ISSUE TITLE
Consider stronger semantics around generated keys

### ISSUE THREAD

**Issue by ajcvickers** [aspnet/EntityFrameworkCore#13575]
Currently we avoid reasoning about store-generated keys in some places because it is possible to add a new entity with an explicit key value set even when the entity type is configured to use generated keys. The two most prominent examples of this are:
* Add behaves differently from Attach. That is, Attach can assume that an entity without a key value set should be Added, while Add cannot assume that an entity with a key value set should be Attached.
* DetectChanges always sets new entities as Added, even if they have a key value set.

Both of these cause confusion.

However, setting key values like this is typically used in special situations--most notably test data and seeding. So adding to the complexity of the mental model and as a result generating confusion in the cases above may not be a good trade-off to make. Instead, if we assume a set store-generated key means that the entity already exists in the database, I believe this would make common things easier to both do and understand.

We could then allow a pattern something like this to temporarily enable setting explicit key values:

```C#
using (context.ChangeTracker.EnableExplitKeys())
{
    context.Add(new Blog { Id = 77 });
}
```

**Comment by sjb-sjb** [comment]
Consideration should be given to allowing fine-grained control over when an entity is attached to a context. The distinction made in EF Core between calling Add and setting State=Added is an example of providing control over whether and when EF automatically / implicitly starts tracking an entity. A systematic review should be done to find all other situations in which EF automatically starts tracking an entity (such as when a navigation property is set on a change-notifying entity) and make sure that there is a straightforward way for the user to disable the automatic start-tracking.

**Comment by ajcvickers** [comment]
@sjb-sjb Please file a separate issue for that--it's not the same thing that this issue is tracking.

**Comment by optiks** [comment]
“However, setting key values like this is typically used in special situations--most notably test data and seeding.“

Not in our case - it’s production code!

We’ve worked around this by forcing the state to Unchanged through the ChangeTracker.Tracked event.

**Comment by ajcvickers** [comment]
@optiks If you're forcing the state to Unchanged, then I think your workaround is doing what I am suggesting we should do by default. That is, if we make this change, then if you set the key value, then we will always assume that this means the state of the entity should be Modified/Unchanged. Conversely, if the key value is not set, then we will always assume that the entity should be Added.

The behavior today follows this except that:
* When Add is called, then all entities in the graph will end up in the Added state because it is _possible_ that key values have been set for some entities even though they should be inserted. I'm saying we don't do this anymore, and if the key values for entities in the graph are set, then we will set the state for those entities to Unchanged.
* When a new entity is detected in an already tracked graph. In this case we currently assume that the entity should be Added, but I am proposing that we instead mark it Unchanged (or possibly Modified) if the key value is set.

**Comment by sjb-sjb** [comment]
@ajcvickers The issue I described here is a restatement of the one I had in #13564.  The point is that EF is automatically attaching the entity when I don't want it to (don't expect it to & can't prevent it). So I would say either we reopen #13564 or leave the comment here.

**Comment by ajcvickers** [comment]
I re-opened #13564 and we will discuss in triage.

**Comment by optiks** [comment]
@sjb-sjb Our workaround may help you (you may need to tweak/make it generic if you don't use int ids):

```
// Subscribe to the Tracked event somewhere, e.g. in the context constructor. You should unsubscribe somewhere too.
ChangeTracker.Tracked += ChangeTracker_Tracked; 
        
private void ChangeTracker_Tracked(object sender, Microsoft.EntityFrameworkCore.ChangeTracking.EntityTrackedEventArgs e)
{
   var entry = e.Entry;
   var entity = entry.Entity;
   if (entry.State == EntityState.Added && IsIdSet(entry))
   {
      entry.State = EntityState.Unchanged;
   }
}

public static bool IsIdSet(EntityEntry entry)
{
   var entity = entry.Entity;
   var metadata = entry.Metadata;
   var primaryKey = metadata.FindPrimaryKey();
   return primaryKey.Properties.All(x => Convert.ToInt32(x.PropertyInfo.GetValue(entity)) > 0);
}

**Comment by sjb-sjb** [comment]
@optiks Will that cause re-entry into the Tracked (or StateChanged) event handler?

**Comment by sjb-sjb** [comment]
@ajcvickers  I agree with your proposal, which I read after suggesting something similar in one of the comments in #13564. 

The only people that could find it problematic would be people who set store-generated keys themselves and, well, they just shouldn't be doing that in the first place!

**Comment by ajcvickers** [comment]
@ajcvickers I disagree your last statement. It's very much a valid thing to do in testing/seeding scenarios. We had many requests to be able to do this in classic EF. However, my point is that we can still support this as an explicit gesture.

The main problem with making this change now is that it's going to be a break for a lot of seeding/tests, so while I am proposing this, I'm not convinced we will be able to do it.

**Comment by optiks** [comment]
@sjb-sjb Unchanged is still tracked, so it shouldn't be re-entrant. If you're subscribing to StateChanged as well, then yes, I'd expect that to fire.

**Comment by sjb-sjb** [comment]
@ajcvickers Good point about seeding. I am currently not seeding in the recommended way, instead I am making internal API calls and then doing a SaveChanges. 

It would make sense to me to go ahead with the changes you propose and then alter HasData so that seeding still works. Within HasData you can move the entities from Unchanged to Added. 

Outside of seeding, it seems like broken behavior to me that EF would Add an entity that already has a valid store-generated key. 

@optiks Thanks for the clarification.

**Comment by optiks** [comment]
@ajcvickers FWIW, I agree that you should be limiting the number of breaking changes. In general, opt-in for behavioural changes is reasonable.

**Comment by optiks** [comment]
@sjb-sjb, I get what you're saying.

Two other workarounds I can think of are:
1. Set the foreign key id property, instead of the navigation property object reference. This may or may not work for you.
2. Pre-emptively attach the foreign key object reference as `Unchanged`. EF Core won't override this state during the Add. 

Another thing to consider is whether disabling [AutoDetectChanges](https://docs.microsoft.com/en-us/ef/ef6/saving/change-tracking/auto-detect-changes) might help in your case?

**Comment by ajcvickers** [comment]
Triage: we agreed on the general idea, but the breaking potential is great. Therefore, we will instead introduce two new methods:
* AddOrAttach
* AddOrUpdate

These methods would work in the way described above when generated keys are being used. That is, if the instance has no key value set, then it will always be Added. If it has a key value set then it will be Unchanged for AddOrAttach or Modified for AddOrUpdate. For non-store generated keys, both will Add everything.

We will make the breaking change in the DetectChanges behavior, but also add a hook to allow the behavior to be defined--see #13564

**Comment by sjb-sjb** [comment]
Just to provide a comment on something that I have found out through a bit more research. In the workaround suggested by @optiks (and I believe this is relevant to #14031 by @rubbiroid as well), setting the state to Unchanged will result in a sequence Detached --> Added --> Unchanged. Some extra work is required to distinguish this from the Detached --> Added --> Unchanged transitions that occur as a result of SaveChanges on an Added entity. The only difference is that in the 'workaround' version we have Id > 0 during Detached --> Added, while in the normal case we have Id < 0.

What we want to see, in effect, is Detached --> Unchanged in the 'workaround' case but we want to keep Detached --> Added --> Unchanged in the 'normal' case. To do this we need to carry information from the Tracked event forward to the StateChanged event. Some locking is needed since the state transition could be triggered from any thread. 

Putting it together we get something like this (replace ".Id" with your primary key):

        private void ContextChangeTracker_Tracked(object sender, EntityTrackedEventArgs e)
        {
            EntityEntry entry = e.Entry;
            EntityState previousState = EntityState.Detached;
            EntityState currentState = entry.State;

            if (previousState == EntityState.Detached && currentState == EntityState.Added && entry.Entity.Id > 0) {
                // Workaround a bug whereby EF can Add an entity that already exists (i.e. that has Id > 0)
                entry.State = EntityState.Unchanged; // triggers StateChanged callback
            } else {
                this.TransitionOccurred(entity, previousState, currentState);
            }
        }

        private void ContextChangeTracker_StateChanged(object sender, EntityStateChangedEventArgs e)
        {
            EntityEntry entry = e.Entry;
            EntityState previousState = e.OldState;
            EntityState currentState = e.NewState;
            Debug.Assert(currentState == entry.State);

            if (previousState == EntityState.Added && currentState == EntityState.Unchanged && entry.Entity.Id > 0 && !this._TrackedEntities.ContainsKey(entry.Entity)) {
                // Second half of the bug fix.
                previousState = EntityState.Detached;
                this.TransitionOccurred(entity, previousState, currentState);
            } else {
                this.TransitionOccurred(entity, previousState, currentState);
            }
        }

        private ConcurrentDictionary<object,bool> _TrackedEntities { get; } = new ConcurrentDictionary<object,bool>(); // not identical to EF tracking.

        private void TransitionOccurred(IEntity entity, EntityState previousState, EntityState subsequentState)
        {
            if (subsequentState == previousState) { return; }
            bool startedTracking = (previousState == EntityState.Detached && subsequentState != EntityState.Detached);
            bool stoppedTracking = (previousState != EntityState.Detached && subsequentState == EntityState.Detached);
          
            lock (this.lockObject) {
                if (startedTracking) {
                     this._TrackedEntities.TryAdd(entity,true);
                } else if (stoppedTracking) {
                     this._TrackedEntities.TryRemove(entity, out bool ignore);
                }
            }
            this.OnTransitionOccurred( entity, previousState, subsequentState);
        }

        virtual protected void OnTransitionOccurred(IEntity entity, EntityState previousState, EntityState subsequentState)
         {
         }

Hope this is useful to the community.

**Comment by ajcvickers** [comment]
DetectChanges bit for 3.0; Backlog for other parts.

**Comment by ajcvickers** [comment]
Split DetectChanges part into #14616

**Comment by sjb-sjb** [comment]
@ajcvickers , @AndriySvyryd : this issue has become more pressing for Sqlite users following the implementation of TPC. 

Since Sqlite does not have named id sequences, the use of TPC with Sqlite typically means using Guid's for the identity field. But for Guid key fields with Sqlite, the key value appears to be generated by EF Core during an Attach operation, and before the Tracked event is called. This differs from autogenerated ints or longs used as identity fields where the value is set into the entity following the save changes. 

Net/net, the workaround above no longer works. Testing the identity field during the Tracked event (arising due to an Attach call) no longer tells us whether or not this is an entity that is already in the database, because the Guid value will have been filled in by EF Core if it was not already set before Attach was called. Consequently we can no longer use this to determine whether the Added state genuinely reflects an entity that needs to be added to the database, vs an entity that already had an Id and was only put into the Added state due to EF's internal logic. Indeed as far as I can tell there is no value associated with the Tracked event args that can be used to test this condition. The only way I know of to work around this is to set a flag in the entity before calling context.Attach(entity) ... clearly not a very desirable situation at all.

**Comment by roji** [comment]
@sjb-sjb you should be able to switch to server-side generation of UUIDs, which would make things very similar to how they work with sequence-drive ints. It seems that there's no nice native way to generate a random UUID in SQLite, but there's at least [this extension](https://github.com/benwebber/sqlite3-uuid) which look pretty good.

Could be related to #30753, which would in general change the default value generation strategy for GUIDs to be server-side.

**Comment by sjb-sjb** [comment]
One way to resolve this would be to store EF-generated guid keys internally, the same way temporary integer ids are, and not store them in the entity field until after the entity has been saved. Treating EF-generated guids like temporary keys would make sense because… they ARE temporary keys, until they are stored in the database whereupon they become permanent keys. The fact that the temporary value is equal to the permanent value is beside the point.

**Comment by sjb-sjb** [comment]
Sure, that’s one way to define it.

**Comment by ajcvickers** [comment]
> Sure, that’s one way to define it.

Importantly, this _is the way it is defined_ for the EF Code. This isn't going to change.

**Comment by sjb-sjb** [comment]
I'm fine with that. I'm just saying that having the Guid's not in the entity field during the Attach would be one way to make a workaround possible here. The breaking impact would be very limited because the few users who do need to access the Guid _during the Attach_ could get it from the shadow property instead. 

Otherwise I don't really know what the workaround could be. I didn't really see how to switch to server-side GUID generation without a large disruption, also if we did that then it would probably have a negative impact on the ability to migrate to other databases in the future.

**Comment by roji** [comment]
> I didn't really see how to switch to server-side GUID generation without a large disruption, also if we did that then it would probably have a negative impact on the ability to migrate to other databases in the future.

What kind of disruption are you thinking of, and why do you think this would impact your ability to migrate to other databases? I'm not aware of other database which don't support server-side generation of GUIDs (SQLite is actually the tricky one).

**Comment by sjb-sjb** [comment]
It's just not an area that I have knowledge in and there seem to be a number of hurdles. The library you referenced says it only works with Linux and OS X, I'm not familiar with installing user functions in Sqlite, and I don't know what I would have to do to enable those functions to generate Id's in EF Core. I also don't know how to configure server-side generation of Guid's in SQL Server or in EF Core with SqlServer. In addition I have also read that there is a possibility of performance issues due to index fragmentation when using Guid's. 

Bear in mind, the problem that I'm really trying to solve is how to use TPC with Sqlite. There is a note on the TPC page saying that autoincremented integer identities don't work with TPC under Sqlite because there is an absence of named sequences in Sqlite. That's why I started investigating using Guid identifiers. But given the challenges with Guid's, I'm wondering whether to look again at the autoincrement question. I understand there is a table in the Sqlite db containing the current autoincrement values, and it seems to me that if I seeded the high-order bits with a different number for each concrete type then the autoincrement approach would work fine. It would also carry-over to SqlServer by just skipping that seeding step. 

I do sort of wish that TPC would work 'out of the box' with Sqlite. Maybe it does and I'm just missing something.
